tests/testtransport.nim

```diff
@@ -8,128 +8,175 @@ import ../libp2p/[connection,
     proc testListener(): Future[bool] {.async, gcsafe.} =
       let ma: MultiAddress = Multiaddress.init("/ip4/0.0.0.0/tcp/0")
       proc connHandler(conn: Connection): Future[void] {.async, gcsafe.} =
+        defer:
+          echo "conn close"
+          await conn.close()
         result = conn.write(cstring("Hello!"), 6)
 
       let transport: TcpTransport = newTransport(TcpTransport)
-      asyncCheck await transport.listen(ma, connHandler)
-      let streamTransport: StreamTransport = await connect(transport.ma)
-      let msg = await streamTransport.read(6)
-      await transport.close()
-      await streamTransport.closeWait()
-
-      result = cast[string](msg) == "Hello!"
-
-    check:
-      waitFor(testListener()) == true
-
-  test "test listener: handle read":
-    proc testListener(): Future[bool] {.async.} =
-      let ma: MultiAddress = Multiaddress.init("/ip4/0.0.0.0/tcp/0")
-      proc connHandler(conn: Connection): Future[void] {.async, gcsafe.} =
-        let msg = await conn.read(6)
-        check cast[string](msg) == "Hello!"
+      defer:
+        echo "transport close"
+        await transport.close()
 
-      let transport: TcpTransport = newTransport(TcpTransport)
       asyncCheck await transport.listen(ma, connHandler)
-      let streamTransport: StreamTransport = await connect(transport.ma)
-      let sent = await streamTransport.write("Hello!", 6)
-      result = sent == 6
-
-    check:
-      waitFor(testListener()) == true
-
-  test "test dialer: handle write":
-    proc testDialer(address: TransportAddress): Future[bool] {.async.} =
-      proc serveClient(server: StreamServer,
-                       transp: StreamTransport) {.async, gcsafe.} =
-        var wstream = newAsyncStreamWriter(transp)
-        await wstream.write("Hello!")
-        await wstream.finish()
-        await wstream.closeWait()
-        await transp.closeWait()
-        server.stop()
-        server.close()
-
-      var server = createStreamServer(address, serveClient, {ReuseAddr})
-      server.start()
-
-      let ma: MultiAddress = MultiAddress.init(server.sock.getLocalAddress())
-      let transport: TcpTransport = newTransport(TcpTransport)
-      let conn = await transport.dial(ma)
-      let msg = await conn.read(6)
-      result = cast[string](msg) == "Hello!"
-
-      server.stop()
-      server.close()
-      await server.join()
-    check waitFor(testDialer(initTAddress("0.0.0.0:0"))) == true
 
-  test "test dialer: handle write":
-    proc testDialer(address: TransportAddress): Future[bool] {.async, gcsafe.} =
-      proc serveClient(server: StreamServer,
-                        transp: StreamTransport) {.async, gcsafe.} =
-        var rstream = newAsyncStreamReader(transp)
-        let msg = await rstream.read(6)
-        check cast[string](msg) == "Hello!"
+      let streamTransport = await connect(transport.ma)
+      defer:
+        echo "streamTransport close"
+        await streamTransport.closeWait()
```

Comment:
@cheatfate I have a issue here that I would like you could check.
For some reason when the defer runs at end of body (backwards obviously)
it works until this call then stops, I checked and there should be no exceptions.

If I call it without `defer` at end works tho. so maybe it's a `defer` issue with async macro.

Reply:
Btw it's very minor issue... I fixed by not using `defer`

Reply:
It looks like Nim compiler problem. To be precise its problem inside exception transformation of closure iterators.

Reply:
https://github.com/nim-lang/Nim/issues/13899

Reply:
Thanks, I kinda expected it would not work but it actually worked most of the time so I started using it...
I guess I will avoid using it or anyway make sure it works.